server/src/main/java/org/elasticsearch/index/engine/SingleDocDirectoryReader.java

```diff
@@ -0,0 +1,275 @@
+            return getDelegate().terms(field);
+        }
+
+        @Override
+        public NumericDocValues getNumericDocValues(String field) throws IOException {
+            return getDelegate().getNumericDocValues(field);
+        }
+
+        @Override
+        public BinaryDocValues getBinaryDocValues(String field) throws IOException {
+            return getDelegate().getBinaryDocValues(field);
+        }
+
+        @Override
+        public SortedDocValues getSortedDocValues(String field) throws IOException {
+            return getDelegate().getSortedDocValues(field);
+        }
+
+        @Override
+        public SortedNumericDocValues getSortedNumericDocValues(String field) throws IOException {
+            return getDelegate().getSortedNumericDocValues(field);
+        }
+
+        @Override
+        public SortedSetDocValues getSortedSetDocValues(String field) throws IOException {
+            return getDelegate().getSortedSetDocValues(field);
+        }
+
+        @Override
+        public NumericDocValues getNormValues(String field) throws IOException {
+            return getDelegate().getNormValues(field);
+        }
+
+        @Override
+        public FieldInfos getFieldInfos() {
+            return getDelegate().getFieldInfos();
+        }
+
+        @Override
+        public Bits getLiveDocs() {
+            return getDelegate().getLiveDocs();
+        }
+
+        @Override
+        public PointValues getPointValues(String field) throws IOException {
+            return getDelegate().getPointValues(field);
+        }
+
+        @Override
+        public void checkIntegrity() throws IOException {
+        }
+
+        @Override
+        public LeafMetaData getMetaData() {
+            return getDelegate().getMetaData();
+        }
+
+        @Override
+        public Fields getTermVectors(int docID) throws IOException {
+            return getDelegate().getTermVectors(docID);
+        }
+
+        @Override
+        public int numDocs() {
+            return 1;
+        }
+
+        @Override
+        public int maxDoc() {
+            return 1;
+        }
+
+        synchronized boolean assertMemorySegmentStatus(boolean loaded) {
+            if (loaded && delegate.get() == null) {
+                assert false : "Expected an in memory segment was loaded; but it wasn't. Please check the reader wrapper implementation";
+            }
+            if (loaded == false && delegate.get() != null) {
+                assert false : "Expected an in memory segment wasn't loaded; but it was. Please check the reader wrapper implementation";
+            }
+            return true;
+        }
+
+        @Override
+        public void document(int docID, StoredFieldVisitor visitor) throws IOException {
+            assert assertMemorySegmentStatus(true);
+            getDelegate().document(docID, visitor);
+        }
+
+        @Override
+        protected void doClose() throws IOException {
+            directory.close();
```

Comment:
Should we also close the delegate (if set)?

Reply:
+1 Could you add a unit test that if you register a close listener on this reader, then it gets called when the reader is called?
And could you also make `getDelegate()` fail once `close()` has been called?

Reply:
++, fixed in https://github.com/elastic/elasticsearch/pull/64504/commits/0dc1a54e1b41bd16b3b4c64743b6b6746c02239c